pkg/operator/csidriveroperator/olmremovalcontroller.go

```diff
@@ -0,0 +1,479 @@
+	// Add informers to the factory now, but the actual event handlers
+	// are added later in CSIDriverOperatorCRController.Run(),
+	// when we're 100% sure the controller is going to start (because it
+	// depends on the platform).
+	// If we added the event handlers now, all events would pile up in the
+	// controller queue, without anything reading it.
+	f = f.WithInformers(
+		clients.OperatorClient.Informer(),
+		// Watch Deployments in any namespace - CSO does not know where the OLM operator is installed
+		clients.KubeInformers.InformersFor("").Apps().V1().Deployments().Informer(),
+		clients.KubeInformers.InformersFor("").Apps().V1().DaemonSets().Informer())
+	f = f.WithNamespaceInformer(
+		clients.KubeInformers.InformersFor("").Core().V1().Namespaces().Informer(),
+		csiOperatorConfig.OLMOptions.CSIDriverNamespace)
+
+	c := &OLMOperatorRemovalController{
+		name:             csiOperatorConfig.ConditionPrefix,
+		operatorClient:   clients.OperatorClient,
+		olmOptions:       csiOperatorConfig.OLMOptions,
+		dynamicClient:    clients.DynamicClient,
+		kubeClient:       clients.KubeClient,
+		nsLister:         clients.KubeInformers.InformersFor("").Core().V1().Namespaces().Lister(),
+		deploymentLister: clients.KubeInformers.InformersFor("").Apps().V1().Deployments().Lister(),
+		daemonSetLister:  clients.KubeInformers.InformersFor("").Apps().V1().DaemonSets().Lister(),
+		eventRecorder:    eventRecorder,
+		factory:          f,
+	}
+	return c
+}
+
+func (c *OLMOperatorRemovalController) Sync(ctx context.Context, syncCtx factory.SyncContext) error {
+	klog.V(4).Infof("OLMOperatorRemovalController.Sync started")
+	defer klog.V(4).Infof("OLMOperatorRemovalController.Sync finished")
+
+	opSpec, _, _, err := c.operatorClient.GetOperatorState()
+	if err != nil {
+		return err
+	}
+	if opSpec.ManagementState != operatorapi.Managed {
+		return nil
+	}
+
+	// 1. Find subscription + namespace
+	subNamespace, subName, csvName, found, err := c.findSubscription(ctx)
+	if err != nil {
+		return err
+	}
+
+	if found {
+		c.olmOperatorNamespace = subNamespace
+		// Delete the subscription, but remember the namespace of the operator
+		// in CR's annotations first - just in case this controller is
+		// restarted after it deletes the Subscription.
+		if err := c.saveOperatorNamespace(subNamespace); err != nil {
+			return err
+		}
+
+		removed, err := c.deleteCSV(ctx, subNamespace, csvName)
+		if err != nil {
+			return err
+		}
+		if !removed {
+			klog.V(4).Infof("OLMOperatorRemovalController.Sync waiting for OLM CSV to disappear")
+			return c.markProgressing("Waiting for OLM CSV to be deleted")
+		}
+
+		removed, err = c.deleteCSV(ctx, subNamespace, csvName)
+		if err != nil {
+			return err
+		}
+		if !removed {
+			klog.V(4).Infof("OLMOperatorRemovalController.Sync waiting for OLM CSV to disappear")
+			return c.markProgressing("Waiting for OLM CSV to be deleted")
+		}
+
+		removed, err = c.deleteSubscription(ctx, subNamespace, subName)
```

Comment:
IMO we should delete the subscription first because otherwise we are racing with OLM operator and it might try to install CSV again if it was deleted, but then we may run into a problem that if we deleted the subscription we wouldn't be able to find the current CSV?